test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java

```diff
@@ -929,62 +929,53 @@ public ClusterHealthStatus waitForRelocation(ClusterHealthStatus status) {
      *                This saves on unneeded searches.
      * @return the actual number of docs seen.
      */
-    public long waitForDocs(final long numDocs, @Nullable final BackgroundIndexer indexer) throws InterruptedException {
+    public long waitForDocs(final long numDocs, @Nullable final BackgroundIndexer indexer) throws Exception {
+        final AtomicLong lastKnownCount = new AtomicLong(-1);
+
         // indexing threads can wait for up to ~1m before retrying when they first try to index into a shard which is not STARTED.
-        return waitForDocs(numDocs, 90, TimeUnit.SECONDS, indexer);
-    }
+        final long maxWaitTimeMs = Math.max(90 * 1000, 200 * numDocs);
 
-    /**
-     * Waits until at least a give number of document is visible for searchers
-     *
-     * @param numDocs         number of documents to wait for
-     * @param maxWaitTime     if not progress have been made during this time, fail the test
-     * @param maxWaitTimeUnit the unit in which maxWaitTime is specified
-     * @param indexer         If supplied it will be first checked for documents indexed.
-     *                        This saves on unneeded searches.
-     * @return the actual number of docs seen.
-     */
-    public long waitForDocs(final long numDocs, int maxWaitTime, TimeUnit maxWaitTimeUnit, @Nullable final BackgroundIndexer indexer)
-        throws InterruptedException {
-        final AtomicLong lastKnownCount = new AtomicLong(-1);
-        long lastStartCount = -1;
-        BooleanSupplier testDocs = () -> {
-            if (indexer != null) {
-                lastKnownCount.set(indexer.totalIndexedDocs());
-            }
-            if (lastKnownCount.get() >= numDocs) {
-                try {
-
-                    long count = client().prepareSearch()
-                        .setTrackTotalHits(true)
-                        .setSize(0)
-                        .setQuery(matchAllQuery())
-                        .get()
-                        .getHits().getTotalHits().value;
-
-                    if (count == lastKnownCount.get()) {
-                        // no progress - try to refresh for the next time
-                        client().admin().indices().prepareRefresh().get();
+        assertBusy(
+            () -> {
+                if (indexer != null) {
+                    lastKnownCount.set(indexer.totalIndexedDocs());
+                }
+
+                // Have we managed to count enough docs yet?
+                if (lastKnownCount.get() < numDocs) {
```

Comment:
The test here was `>=` before and I couldn't see why that would be correct. Please point out what I've missed.

Reply:
Do we need any of this state between assertions anymore if we are no longer doing a loop? Shouldn't the assertion should be visible docs >= numDocs?

Reply:
I don't think we need this state across assertions anymore? We can just do a normal query/assert num docs, no need to check for progress.

Reply:
The JavaDoc says this about the `indexer` argument (which is optional only in the `ESIntegTestCase` implementation:

> If supplied it will be first checked for documents indexed. This saves on unneeded searches.

So it would make sense if the code checked the indexer's number of documents first, and only if it's lower that the desired count would it then perform a search (presumably hoping that more documents have been indexed?). I don't understand the code would check the indexer's number of docs, and only if that number is `>= numDocs` would it perform the search (which is just a search to find the total number of hits and doesn't actually return any documents). Unless the code needs the indexer to think it has attempted to index the desired number of docs first?

Reply:
Aha, the indexed vs searchable difference is what I was missing. That makes much more sense now.

I had kept the document count because the method returned it, but actually none of the call sites use the number, so I've changed it to return `void` instead.